rkt/trust.go

```diff
@@ -0,0 +1,230 @@
+	if len(args) != 1 {
+		printCommandUsageByName(cmdTrustName)
+		return 1
+	}
+
+	pkl, err := getPubKeyLocation(flagPrefix, args)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error determining key location: %v\n", err)
+		return 1
+	}
+
+	pk, err := getPubKey(pkl)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error accessing key: %v\n", err)
+		return 1
+	}
+	defer pk.Close()
+
+	accepted, err := reviewKey(pk, flagAccept)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error reviewing key: %v\n", err)
+	}
+
+	if !accepted {
+		fmt.Fprintf(os.Stderr, "Not trusting %q.\n", pkl)
+		return 0
+	}
+
+	fmt.Printf("Trusting %q for prefix %q.\n", pkl, flagPrefix)
+
+	if err := addPubKey(flagPrefix, pk); err != nil {
+		fmt.Fprintf(os.Stderr, "Error adding key: %v\n", err)
+		return 1
+	}
+
+	return 0
+}
+
+// getPubKeyLocation either returns the location supplied in argv or discovers one @ prefix
+func getPubKeyLocation(prefix string, args []string) (string, error) {
+	if len(args) == 1 {
+		return args[0], nil
+	}
+
+	if len(args) > 1 {
+		return "", fmt.Errorf("too many arguments")
+	}
+
+	if prefix == "" {
+		return "", fmt.Errorf("a key or --prefix required")
+	}
+
+	// TODO(vc): discover location from --prefix and return it here?
+
+	return "", nil
+}
+
+// getPubKey retrieves a public key (if remote), and verifies it's a gpg key
+func getPubKey(location string) (*os.File, error) {
+	u, err := url.Parse(location)
+	if err != nil {
+		return nil, err
+	}
+
+	switch u.Scheme {
+	case "":
+		return os.Open(location)
+	case "http":
+		fallthrough
+	case "https":
+		return downloadKey(u.String())
+	default:
+	}
+	return nil, fmt.Errorf("only http and https urls supported")
+}
+
+// downloadKey retrieves the file, storing it in a deleted tempfile
+func downloadKey(url string) (*os.File, error) {
+	tf, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("error creating tempfile: %v", err)
+	}
+	os.Remove(tf.Name()) // no need to keep the tempfile around
+
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("error getting key: %v", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad HTTP status code: %d", res.StatusCode)
+	}
+
+	if _, err := io.Copy(tf, res.Body); err != nil {
+		tf.Close()
+		return nil, fmt.Errorf("error copying key: %v", err)
+	}
+
+	tf.Seek(0, os.SEEK_SET)
+
+	return tf, nil
+}
+
+func fingerToString(fpr [20]byte) string {
+	str := ""
+	for i, b := range fpr {
+		if i > 0 && i%2 == 0 {
+			str += " "
+			if i == 10 {
+				str += " "
+			}
+		}
+		str += strings.ToUpper(fmt.Sprintf("%.2x", b))
+	}
+	return str
+}
+
+// reviewKey shows the key summary and conditionally asks the user to accept it
+func reviewKey(key *os.File, forceAccept bool) (bool, error) {
+	result := false
+
+	defer key.Seek(0, os.SEEK_SET)
+
+	kr, err := openpgp.ReadArmoredKeyRing(key)
+	if err != nil {
+		return false, fmt.Errorf("error reading key: %v", err)
+	}
+
```

Comment:
Yes, ignore me, I'm contradicting myself with the idea that this shouldn't really be scripted